hw/drivers/pwm/pwm_nrf52/src/pwm_nrf52.c

```diff
@@ -177,9 +249,10 @@ nrf52_pwm_close(struct os_dev *odev)
 
     dev = (struct pwm_dev *) odev;
     inst_id = dev->pwm_instance_id;
+    assert(instances[inst_id].in_use);
```

Comment:
I think probably all these asserts needs to be just return codes. My understanding is that asserts are for system startup where theres no one to return to or no reason to keep going on failure, not runtime errors especially when the function already allows a return code

Reply:
I think an assert is appropriate for certain failure conditions:
* software bugs (i.e., logical contradiction)
* unlikely hardware failure

In the case of a software bug, my view is that it is best to terminate as soon as possible.  This makes it easier to identify and fix the bug, and generally there is no reasonable recovery mechanism.  For "safety critical" systems, presumably there are hardware recovery mechanisms or redundancies that protect against a software crash.

I am not familiar with this PWM driver, but it looks like this `assert` will only fail if there is a firmware bug.  There are similar asserts that check this same condition elsewhere in this file.

Anyway, I'm sure opinions vary, so that is just my take.

Reply:
> I think probably all these asserts needs to be just return codes.

Agreed, on this case it makes perfect sense to have error codes. 

IMHO, I'd say use error codes on recoverable situations, use asserts on unrecoverable situations (i.e. checking NULL pointers, array lengths, etc). I've seen here code produced under this practice, I think if everyone agrees, it would be